File: 2019/intcode.ts

type OpResult = {
  state: number[],
  ip: number,
  rp: number,
};

enum Mode {
  Position = 0,
  Immediate = 1,
  Relative = 2,
}

class Op {
  op: number;
  ip: number;
  rp: number;
  params: number[];
  modes: number[];

  constructor(prev: number[], ip: number, rp: number) {
    const code = prev[ip];
    this.ip = ip;
    this.rp = rp;
    this.op = code % 100;
    this.modes = Math.trunc(code / 100).toString().split("").map(Number).reverse();
    this.params = prev.slice(ip + 1, ip + 1 + this.paramCount());
  }

  paramCount() {
    return {
      1: 3,
      2: 3,
      3: 1,
      4: 1,
      5: 2,
      6: 2,
      7: 3,
      8: 3,
      9: 1,
      99: 0,
    }[this.op];
  }

  value(position: number, state: number[], writeMode = false): number {
    const mode = this.modes[position] || Mode.Position;

    if (mode === Mode.Position) {
      if (writeMode) {
        return this.params[position];
      }
      return state[this.params[position]] || 0;
    } else if (mode === Mode.Immediate) {
      return this.params[position];
    } else if (mode === Mode.Relative) {
      if (writeMode) {
        return this.params[position] + this.rp;
      }
      return state[this.params[position] + this.rp] || 0;
    }
  }

  nextStep() {
    return this.ip + this.paramCount() + 1;
  }

  async execute(state: number[], inputs: () => Promise<number>, outputs: number[]): Promise<OpResult> {
    const updated = (memory: number[], idx: number | number, value: number | number): number[] => {
      const copy = memory.slice();
      copy[this.value(idx, copy, true)] = value;

      return copy;
    };

    const result = {
      state,
      ip: this.nextStep(),
      rp: this.rp,
    };

    switch (this.op) {
      case 1:
        return {
          ...result,
          state: updated(state, 2, this.value(0, state) + this.value(1, state)),
        };
      case 2:
        return {
          ...result,
          state: updated(state, 2, this.value(0, state) * this.value(1, state)),
        };
      case 3:
        return {
          ...result,
          state: updated(state, 0, await inputs()),
        };
      case 4:
        outputs.push(this.value(0, state));
        return result;
      case 5:
        if (this.value(0, state) !== 0) {
          return { ...result, ip: this.value(1, state) };
        }
        return result;
      case 6:
        if (this.value(0, state) === 0) {
          return { ...result, ip: this.value(1, state) };
        }
        return result;
      case 7:
        return {
          ...result,
          state: updated(state, 2, this.value(0, state) < this.value(1, state) ? 1 : 0),
        };
      case 8:
        return {
          ...result,
          state: updated(state, 2, this.value(0, state) === this.value(1, state) ? 1 : 0),
        };
      case 9:
        return {
          ...result,
          rp: result.rp + this.value(0, state),
        };
      case 99:
        return {
          ...result,
          ip: this.ip,
        };
      default:
        throw new Error("Unexpected op" + this.op);
    }
  }
}

export default class Program {
  state: number[];
  ip = 0;
  rp = 0;
  inputs: number[];
  inputFn: () => Promise<number>;
  process: AsyncGenerator<number>;

  constructor(program: number[], inputs: number[] = []) {
    this.state = program.slice(0);
    this.inputs = inputs.slice(0);
    this.process = this.run();
    this.inputFn = () => Promise.resolve(this.inputs.shift());
  }

  addInput(i: number) {
    this.inputs.push(i);
    return this;
  }

  dup(): Program {
    const p = new Program(this.state.slice(0), this.inputs)

    p.ip = this.ip;
    p.rp = this.rp;

    return p;
  }

  async next(): Promise<IteratorResult<number, any>> {
    return this.process.next();
  }

  async nextVal(): Promise<number> {
    return (await this.process.next()).value;
  }

  private async * run(): AsyncGenerator<number> {
    while (this.state[this.ip] !== 99) {
      const outputs = [];
      const result = await (new Op(this.state, this.ip, this.rp).execute(this.state, this.inputFn, outputs));
      this.ip = result.ip;
      this.rp = result.rp;
      this.state = result.state;

      while (outputs.length > 0) {
        yield outputs.shift();
      }
    }
  }
}

File: intcode-kotlin/src/main/kotlin/intcode.kt
import java.io.File

data class OpResult(val state: List<Long>, val ip: Int, val rp: Int, val outputs: List<Long> = emptyList())
enum class Mode {
  Position,
  Immediate,
  Relative
}

class Op(private val rp: Int, private val ip: Int, prev: List<Long>) {
  private val op: Int
  private val modes: List<Int>
  private val params: List<Long>
  init {
    val code = prev[ip]
    op = (code % 100).toInt()
    modes = (code / 100).toString(radix = 10).map { Character.getNumericValue(it) }.reversed()
    params = prev.subList(ip + 1, ip + 1 + opParamCount(op))
  }

  private fun nextStep(): Int = ip + paramCount() + 1
  private fun paramCount(): Int = opParamCount(op)
  private fun value(position: Int, state: List<Long>, writeMode: Boolean = false): Long =
    when (Mode.values()[modes.getOrElse(position) { 0 }]) {
      Mode.Position -> if (writeMode) params[position]
      else state.getOrElse(params[position].toInt()) { 0 }
      Mode.Immediate -> params[position]
      Mode.Relative -> if (writeMode) params[position] + rp
      else state.getOrElse(params[position].toInt() + rp) { 0 }
    }

  fun execute(state: List<Long>, inputFn: () -> Long): OpResult {
    val default = OpResult(state, nextStep(), rp)

    fun updated(memory: List<Long>, idx: Int, value: Long): List<Long> =
      memory.toMutableList().apply {
        // Dynamically resize memory
        val address = value(idx, this, true).toInt()
        if (address >= memory.size) {
          (memory.size..(address + 1)).forEach { add(it, 0L) }
        }
        set(address, value)
      }.toList()

    return when (op) {
      1 -> default.copy(state = updated(state, 2, value(0, state) + value(1, state)))
      2 -> default.copy(state = updated(state, 2, value(0, state) * value(1, state)))
      3 -> default.copy(state = updated(state, 0, inputFn()))
      4 -> default.copy(outputs = listOf(value(0, state)))
      5 -> if (value(0, state) != 0L) default.copy(ip = value(1, state).toInt()) else default
      6 -> if (value(0, state) == 0L) default.copy(ip = value(1, state).toInt()) else default
      7 -> default.copy(state = updated(state, 2, if (value(0, state) < value(1, state)) 1 else 0))
      8 -> default.copy(state = updated(state, 2, if (value(0, state) == value(1, state)) 1 else 0))
      9 -> default.copy(rp = default.rp + value(0, state).toInt())
      99 -> default.copy(ip = ip)
      else -> throw Exception("What is $op?")
    }
  }

  companion object {
    fun opParamCount(op: Int): Int =
      mapOf(1 to 3, 2 to 3, 3 to 1, 4 to 1, 5 to 2, 6 to 2, 7 to 3, 8 to 3, 9 to 1, 99 to 0)[op]!!
  }
}

class Program(private var state: List<Long>, inputSeed: List<Long> = emptyList()) {
  private var ip = 0
  private var rp = 0
  private var inputFn = fun(): Long = inputs.removeFirst()
  private val inputs: MutableList<Long> = inputSeed.toMutableList()

  fun valueAt(address: Int) = state[address]

  fun addInput(new: List<Long>) {
    this.inputs.addAll(new)
  }

  fun clone(): Program = Program(state, inputs).also {
    it.rp = this.rp
    it.ip = this.ip
  }

  fun start(): Sequence<Long> {
    return sequence {
      while (state[ip] != 99L) {
        Op(rp, ip, state).execute(state, inputFn).let { result ->
          ip = result.ip
          rp = result.rp
          state = result.state

          result.outputs.forEach { yield(it) }
        }
      }
    }
  }

  companion object {
    fun fromFile(file: String, inputs: List<Long> = emptyList()): Program {
      return Program(
        File(file).readText().split(",").map(String::trim).map(String::toLong),
        inputs
      )
    }
  }
}


